File: ios/App/App/IsolatedModules/IsolatedModules.swift
//
//  IsolatedModules.swift
//  App
//
//  Created by Julia Samol on 08.09.22.
//

import Foundation
import Capacitor
import CryptoKit

@objc(IsolatedModules)
public class IsolatedModules: CAPPlugin {
    private let fileExplorer: FileExplorer = .shared
    private lazy var jsEvaluator: JSEvaluator = .init(fileExplorer: fileExplorer)
    
    @objc func previewDynamicModule(_ call: CAPPluginCall) {
        call.assertReceived(forMethod: "previewDynamicModule", requiredParams: Param.PATH, Param.DIRECTORY)
        
        do {
            guard let path = call.path, let directory = call.directory else {
                throw Error.invalidData
            }
            
            Task {
                do {
                    let module = try fileExplorer.loadPreviewModule(atPath: path, locatedIn: directory)
                    let manifest = try fileExplorer.readModuleManifest(.preview(module))
                    let moduleJSON = try await jsEvaluator.evaluatePreviewModule(
                        .preview(module),
                        ignoringProtocols: call.ignoreProtocols
                    )
                    
                    call.resolve([
                        "module": moduleJSON,
                        "manifest": try JSONSerialization.jsonObject(with: manifest)
                    ])
                } catch {
                    call.reject("Error: \(error)")
                }
            }
        } catch {
            call.reject("Error: \(error)")
        }
    }
    
    @objc func verifyDynamicModule(_ call: CAPPluginCall) {
        call.assertReceived(forMethod: "verifyDynamicModule", requiredParams: Param.PATH, Param.DIRECTORY)
        
        do {
            guard let path = call.path, let directory = call.directory else {
                throw Error.invalidData
            }
            
            Task {
                do {
                    let module = try fileExplorer.loadPreviewModule(atPath: path, locatedIn: directory)
                    let manifest = try fileExplorer.readModuleManifest(.preview(module))
                    let files = try fileExplorer.readModuleFiles(.preview(module))
                    
                    let message = (files + [manifest]).reduce(Data()) { acc, next in acc + next }

                    let jsonDecoder = JSONDecoder()
                    let publicKey = try jsonDecoder.decode(ModuleManifest.self, from: manifest).publicKey

                    let verified = try Curve25519.Signing.PublicKey(rawRepresentation: publicKey.asBytes()).isValidSignature(module.signature, for: message)
                    
                    call.resolve([
                        "verified": verified
                    ])
                } catch {
                    call.reject("Error: \(error)")
                }
            }
        } catch {
            call.reject("Error: \(error)")
        }
    }
    
    @objc func registerDynamicModule(_ call: CAPPluginCall) {
        call.assertReceived(forMethod: "registerDynamicModule", requiredParams: Param.IDENTIFIER, Param.PROTOCOL_IDENTIFIERS)
        
        do {
            guard let identifier = call.identifier, let protocolIdentifiers = call.protocolIdentifiers else {
                throw Error.invalidData
            }
            
            Task {
                do {
                    let module = try fileExplorer.loadInstalledModule(identifier)
                    await jsEvaluator.registerModule(.installed(module), forProtocols: protocolIdentifiers)
                    
                    call.resolve()
                } catch {
                    call.reject("Error: \(error)")
                }
            }
        } catch {
            call.reject("Error: \(error)")
        }
    }
    
    @objc func readDynamicModule(_ call: CAPPluginCall) {
        call.assertReceived(forMethod: "readDynamicModule", requiredParams: Param.IDENTIFIER)
        
        do {
            guard let identifier = call.identifier else {
                throw Error.invalidData
            }
            
            Task {
                do {
                    let module = try fileExplorer.loadInstalledModule(identifier)
                    let manifest = try fileExplorer.readModuleManifest(.installed(module))
                    
                    call.resolve([
                        "manifest": try JSONSerialization.jsonObject(with: manifest),
                        "installedAt": "\(module.installedAt)"
                    ])
                } catch {
                    call.reject("Error: \(error)")
                }
            }
        } catch {
            call.reject("Error: \(error)")
        }
    }
    
    @objc func removeDynamicModules(_ call: CAPPluginCall) {
        Task {
            do {
                if let identifiers = call.identifiers {
                    try fileExplorer.removeModules(identifiers)
                    await jsEvaluator.deregisterModules(identifiers)
                } else {
                    try fileExplorer.removeAllModules()
                    await jsEvaluator.deregisterAllModules()
                }
                call.resolve()
            } catch {
                call.reject("Error: \(error)")
            }
        }
    }
    
    @objc func readAssetModule(_ call: CAPPluginCall) {
        call.assertReceived(forMethod: "readAssetModule", requiredParams: Param.IDENTIFIER)
        
        do {
            guard let identifier = call.identifier else {
                throw Error.invalidData
            }
            
            Task {
                do {
                    let module = try fileExplorer.loadAssetModule(identifier)
                    let manifest = try fileExplorer.readModuleManifest(.asset(module))
                    
                    call.resolve([
                        "manifest": try JSONSerialization.jsonObject(with: manifest)
                    ])
                } catch {
                    call.reject("Error: \(error)")
                }
            }
        } catch {
            call.reject("Error: \(error)")
        }
    }
    
    @objc func loadAllModules(_ call: CAPPluginCall) {
        Task {
            do {
                let protocolType = call.protocolType
                let modules: [JSModule] = try fileExplorer.loadAssetModules().map({ .asset($0) }) + (try fileExplorer.loadInstalledModules().map({ .installed($0) }))
                
                call.resolve(
                    try await jsEvaluator.evaluateLoadModules(
                        modules,
                        for: protocolType,
                        ignoringProtocols: call.ignoreProtocols
                    )
                )
            } catch {
                call.reject("Error: \(error)")
            }
        }
    }
    
    @objc func callMethod(_ call: CAPPluginCall) {
        call.assertReceived(forMethod: "callMethod", requiredParams: Param.TARGET, Param.METHOD)
        
        do {
            guard let target = call.target, let method = call.method else {
                throw Error.invalidData
            }
            
            Task {
                do {
                    switch target {
                    case .offlineProtocol:
                        call.assertReceived(forMethod: "callMethod", requiredParams: Param.PROTOCOL_IDENTIFIER)
                        
                        guard let protocolIdentifier = call.protocolIdentifier else {
                            throw Error.invalidData
                        }
                        
                        let args = call.args
                        
                        call.resolve(
                            try await jsEvaluator.evaluateCallOfflineProtocolMethod(method, ofProtocol: protocolIdentifier, withArgs: args)
                        )
                    case .onlineProtocol:
                        call.assertReceived(forMethod: "callMethod", requiredParams: Param.PROTOCOL_IDENTIFIER)
                        
                        guard let protocolIdentifier = call.protocolIdentifier else {
                            throw Error.invalidData
                        }
                        
                        let args = call.args
                        let networkID = call.networkID
                        
                        call.resolve(
                            try await jsEvaluator.evaluateCallOnlineProtocolMethod(method, ofProtocol: protocolIdentifier, onNetwork: networkID, withArgs: args)
                        )
                    case .blockExplorer:
                        call.assertReceived(forMethod: "callMethod", requiredParams: Param.PROTOCOL_IDENTIFIER)
                        
                        guard let protocolIdentifier = call.protocolIdentifier else {
                            throw Error.invalidData
                        }
                        
                        let args = call.args
                        let networkID = call.networkID
                        
                        call.resolve(
                            try await jsEvaluator.evaluateCallBlockExplorerMethod(method, ofProtocol: protocolIdentifier, onNetwork: networkID, withArgs: args)
                        )
                    case .v3SerializerCompanion:
                        call.assertReceived(forMethod: "callMethod", requiredParams: Param.MODULE_IDENTIFIER)
                        
                        guard let moduleIdentifier = call.moduleIdentifier else {
                            throw Error.invalidData
                        }
                        
                        let args = call.args
                        
                        call.resolve(
                            try await jsEvaluator.evaluateCallV3SerializerCompanionMethod(method, ofModule: moduleIdentifier, withArgs: args)
                        )
                    }
                } catch {
                    call.reject("Error: \(error)")
                }
            }
        } catch {
            call.reject("Error: \(error)")
        }
    }
    
    @objc func batchCallMethod(_ call: CAPPluginCall) {
        call.assertReceived(forMethod: "batchCallMethod", requiredParams: Param.OPTIONS)
        
        do {
            guard let options = call.options else {
                throw Error.invalidData
            }
            
            Task {
                do {
                    let values = try await self.jsEvaluator.singleRun { runRef in
                        try await options.asyncMap { jsValue in
                            do {
                                guard let jsObject = jsValue as? JSObject else {
                                    throw Error.invalidData
                                }
                                
                                call.assertReceived(in: jsObject, forMethod: "batchCallMethod", requiredParams: Param.TARGET, Param.METHOD)
                                
                                guard let targetRaw = jsObject[Param.TARGET] as? String,
                                      let target = JSCallMethodTarget.init(rawValue: targetRaw),
                                      let method = jsObject[Param.METHOD] as? String else {
                                    throw Error.invalidData
                                }
                                
                                let args = jsObject[Param.ARGS] as? JSArray
                                let protocolIdentifier = jsObject[Param.PROTOCOL_IDENTIFIER] as? String
                                let networkID = jsObject[Param.NETWORK_ID] as? String
                                let moduleIdentifier = jsObject[Param.MODULE_IDENTIFIER] as? String
                                
                                let value = try await {
                                    switch target {
                                    case .offlineProtocol:
                                        call.assertReceived(in: jsObject, forMethod: "batchCallMethod", requiredParams: Param.PROTOCOL_IDENTIFIER)
                                        
                                        guard let protocolIdentifier = protocolIdentifier else {
                                            throw Error.invalidData
                                        }
                                        
                                        return try await self.jsEvaluator.evaluateCallOfflineProtocolMethod(
                                            method,
                                            ofProtocol: protocolIdentifier,
                                            withArgs: args,
                                            runRef: runRef
                                        )
                                    case .onlineProtocol:
                                        call.assertReceived(in: jsObject, forMethod: "batchCallMethod", requiredParams: Param.PROTOCOL_IDENTIFIER)
                                        
                                        guard let protocolIdentifier = protocolIdentifier else {
                                            throw Error.invalidData
                                        }
                                        
                                        return try await self.jsEvaluator.evaluateCallOnlineProtocolMethod(
                                            method,
                                            ofProtocol: protocolIdentifier,
                                            onNetwork: networkID,
                                            withArgs: args,
                                            runRef: runRef
                                        )
                                    case .blockExplorer:
                                        call.assertReceived(in: jsObject, forMethod: "batchCallMethod", requiredParams: Param.PROTOCOL_IDENTIFIER)
                                        
                                        guard let protocolIdentifier = protocolIdentifier else {
                                            throw Error.invalidData
                                        }
                                        
                                        return try await self.jsEvaluator.evaluateCallBlockExplorerMethod(
                                            method,
                                            ofProtocol: protocolIdentifier,
                                            onNetwork: networkID,
                                            withArgs: args,
                                            runRef: runRef
                                        )
                                    case .v3SerializerCompanion:
                                        call.assertReceived(in: jsObject, forMethod: "batchCallMethod", requiredParams: Param.MODULE_IDENTIFIER)
                                        
                                        guard let moduleIdentifier = moduleIdentifier else {
                                            throw Error.invalidData
                                        }
                                        
                                        return try await self.jsEvaluator.evaluateCallV3SerializerCompanionMethod(
                                            method,
                                            ofModule: moduleIdentifier,
                                            withArgs: args,
                                            runRef: runRef
                                        )
                                    }
                                }()
                                
                                return [
                                    "type": "success",
                                    "value": value["value"]
                                ]
                            } catch {
                                return [
                                    "type": "error",
                                    "error": "\(error)"
                                ]
                            }
                        }
                    }
                    
                    call.resolve([
                        "values": values
                    ])
                } catch {
                    call.reject("Error: \(error)")
                }
            }
        } catch {
            call.reject("Error: \(error)")
        }
    }
    
    struct Param {
        static let PATH = "path"
        static let DIRECTORY = "directory"
        static let IDENTIFIER = "identifier"
        static let IDENTIFIERS = "identifiers"
        static let PROTOCOL_IDENTIFIERS = "protocolIdentifiers"
        static let PROTOCOL_TYPE = "protocolType"
        static let IGNORE_PROTOCOLS = "ignoreProtocols"
        static let TARGET = "target"
        static let METHOD = "method"
        static let ARGS = "args"
        static let PROTOCOL_IDENTIFIER = "protocolIdentifier"
        static let MODULE_IDENTIFIER = "moduleIdentifier"
        static let NETWORK_ID = "networkId"
        static let OPTIONS = "options"
    }
    
    enum Error: Swift.Error {
        case invalidData
    }
}

private extension CAPPluginCall {
    var path: String? { return getString(IsolatedModules.Param.PATH) }
    
    var directory: Directory? {
        guard let directory = getString(IsolatedModules.Param.DIRECTORY) else { return nil }
        return .init(rawValue: directory)
    }
    
    var identifier: String? { return getString(IsolatedModules.Param.IDENTIFIER) }
    var identifiers: [String]? {
        return getArray(IsolatedModules.Param.IDENTIFIERS)?.compactMap {
            if let string = $0 as? String {
                return string
            } else {
                return nil
            }
        }
    }
    var protocolIdentifiers: [String]? {
        return getArray(IsolatedModules.Param.PROTOCOL_IDENTIFIERS)?.compactMap {
            if let string = $0 as? String {
                return string
            } else {
                return nil
            }
        }
    }
    
    var protocolType: JSProtocolType? {
        guard let protocolType = getString(IsolatedModules.Param.PROTOCOL_TYPE) else { return nil }
        return .init(rawValue: protocolType)
    }
    
    var ignoreProtocols: [String]? {
        return getArray(IsolatedModules.Param.IGNORE_PROTOCOLS)?.compactMap {
            if let string = $0 as? String {
                return string
            } else {
                return nil
            }
        }
    }
    
    var target: JSCallMethodTarget? {
        guard let target = getString(IsolatedModules.Param.TARGET) else { return nil }
        return .init(rawValue: target)
    }
    
    var method: String? { return getString(IsolatedModules.Param.METHOD) }
    var args: JSArray? { return getArray(IsolatedModules.Param.ARGS) }
    
    var protocolIdentifier: String? { return getString(IsolatedModules.Param.PROTOCOL_IDENTIFIER) }
    var moduleIdentifier: String? { return getString(IsolatedModules.Param.MODULE_IDENTIFIER) }
    
    var networkID: String? { return getString(IsolatedModules.Param.NETWORK_ID) }
    
    var options: JSArray? { return getArray(IsolatedModules.Param.OPTIONS) }
}

File: android/app/src/main/java/it/airgap/wallet/plugin/isolatedmodules/IsolatedModules.kt
package it.airgap.wallet.plugin.isolatedmodules

import androidx.lifecycle.lifecycleScope
import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import it.airgap.wallet.plugin.isolatedmodules.js.JSCallMethodTarget
import it.airgap.wallet.plugin.isolatedmodules.js.JSEvaluator
import it.airgap.wallet.plugin.isolatedmodules.js.JSProtocolType
import it.airgap.wallet.util.*
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.launch
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer
import org.json.JSONObject

@CapacitorPlugin
class IsolatedModules : Plugin() {
    private val jsEvaluator: Deferred<JSEvaluator> = ExecutableDeferred { JSEvaluator(context, fileExplorer) }
    private val fileExplorer: FileExplorer by lazy { FileExplorer(context) }

    @PluginMethod
    fun previewDynamicModule(call: PluginCall) {
        call.executeCatching {
            assertReceived(Param.PATH, Param.DIRECTORY)

            activity.lifecycleScope.launch {
                executeCatching {
                    val module = fileExplorer.loadPreviewModule(path, directory)
                    val manifest = fileExplorer.readModuleManifest(module)
                    val moduleJson = jsEvaluator.await().evaluatePreviewModule(module)

                    resolve(
                        JSObject(
                            """
                                {
                                    "module": $moduleJson,
                                    "manifest": ${JSObject(manifest.decodeToString())}
                                }
                            """.trimIndent()
                        )
                    )
                }
            }
        }
    }

    @PluginMethod
    fun verifyDynamicModule(call: PluginCall) {
        call.executeCatching {
            assertReceived(Param.PATH, Param.DIRECTORY)

            activity.lifecycleScope.launch {
                executeCatching {
                    val module = fileExplorer.loadPreviewModule(path, directory)
                    val manifest = fileExplorer.readModuleManifest(module)
                    val files = fileExplorer.readModuleFiles(module)

                    val message = (files + manifest).reduce(ByteArray::plus)
                    val publicKey = JSObject(manifest.decodeToString()).getString("publicKey")!!

                    val verified = Ed25519Signer()
                        .apply {
                            init(false, Ed25519PublicKeyParameters(publicKey.asByteArray()))
                            update(message, 0, message.size)
                        }
                        .verifySignature(module.signature.asByteArray())

                    resolve(
                        JSObject(
                            """
                                {
                                    "verified": $verified
                                }
                            """.trimIndent()
                        )
                    )
                }
            }
        }
    }

    @PluginMethod
    fun registerDynamicModule(call: PluginCall) {
        call.executeCatching {
            assertReceived(Param.IDENTIFIER, Param.PROTOCOL_IDENTIFIERS)

            activity.lifecycleScope.launch {
                executeCatching {
                    val module = fileExplorer.loadInstalledModule(identifier)
                    jsEvaluator.await().registerModule(module, protocolIdentifiers)

                    resolve()
                }
            }
        }
    }

    @PluginMethod
    fun readDynamicModule(call: PluginCall) {
        call.executeCatching {
            assertReceived(Param.IDENTIFIER)

            activity.lifecycleScope.launch {
                executeCatching {
                    val module = fileExplorer.loadInstalledModule(identifier)
                    val manifest = fileExplorer.readModuleManifest(module)

                    resolve(
                        JSObject(
                            """
                                {
                                    "manifest": ${JSObject(manifest.decodeToString())},
                                    "installedAt": "${module.installedAt}"
                                }
                            """.trimIndent()
                        )
                    )
                }
            }
        }
    }

    @PluginMethod
    fun removeDynamicModules(call: PluginCall) {
        activity.lifecycleScope.launch {
            call.executeCatching {
                val jsEvaluator = jsEvaluator.await()

                identifiers?.let {
                    fileExplorer.removeInstalledModules(it)
                    jsEvaluator.deregisterModules(it)
                } ?: run {
                    fileExplorer.removeAllInstalledModules()
                    jsEvaluator.deregisterAllModules()
                }

                resolve()
            }
        }
    }

    @PluginMethod
    fun readAssetModule(call: PluginCall) {
        call.executeCatching {
            assertReceived(Param.IDENTIFIER)

            activity.lifecycleScope.launch {
                executeCatching {
                    val module = fileExplorer.loadAssetModule(identifier)
                    val manifest = fileExplorer.readModuleManifest(module)

                    resolve(
                        JSObject(
                            """
                                {
                                    "manifest": ${JSObject(manifest.decodeToString())}
                                }
                            """.trimIndent()
                        )
                    )
                }
            }
        }
    }

    @PluginMethod
    fun loadAllModules(call: PluginCall) {
        activity.lifecycleScope.launch {
            call.executeCatching {
                val modules = fileExplorer.loadAssetModules() + fileExplorer.loadInstalledModules()

                resolve(jsEvaluator.await().evaluateLoadModules(modules, protocolType, ignoreProtocols))
            }
        }
    }

    @PluginMethod
    fun callMethod(call: PluginCall) {
        call.executeCatching {
            assertReceived(Param.TARGET, Param.METHOD)

            activity.lifecycleScope.launch {
                executeCatching {
                    val value = when (target) {
                        JSCallMethodTarget.OfflineProtocol -> {
                            assertReceived(Param.PROTOCOL_IDENTIFIER)
                            jsEvaluator.await().evaluateCallOfflineProtocolMethod(method, args, protocolIdentifier)
                        }
                        JSCallMethodTarget.OnlineProtocol -> {
                            assertReceived(Param.PROTOCOL_IDENTIFIER)
                            jsEvaluator.await().evaluateCallOnlineProtocolMethod(method, args, protocolIdentifier, networkId)
                        }
                        JSCallMethodTarget.BlockExplorer -> {
                            assertReceived(Param.PROTOCOL_IDENTIFIER)
                            jsEvaluator.await().evaluateCallBlockExplorerMethod(method, args, protocolIdentifier, networkId)
                        }
                        JSCallMethodTarget.V3SerializerCompanion -> {
                            assertReceived(Param.MODULE_IDENTIFIER)
                            jsEvaluator.await().evaluateCallV3SerializerCompanionMethod(method, args, moduleIdentifier)
                        }
                    }
                    resolve(value)
                }
            }
        }
    }

    @PluginMethod
    fun batchCallMethod(call: PluginCall) {
        call.executeCatching {
            assertReceived(Param.OPTIONS)

            activity.lifecycleScope.launch {
                executeCatching {
                    val jsEvaluator = jsEvaluator.await()

                    val values = jsEvaluator.singleRun { runRef ->
                        options.toList<JSONObject>().asyncMap { jsonObject ->
                            try {
                                val jsObject = JSObject(jsonObject.toString())
                                assertReceivedIn(jsObject, Param.TARGET, Param.METHOD)

                                val target = jsObject.getString(Param.TARGET)?.let { JSCallMethodTarget.fromString(it) }!!
                                val method = jsObject.getString(Param.METHOD)!!
                                val args = if (jsObject.has(Param.ARGS)) JSArray(jsObject.getJSONArray(Param.ARGS).toString()) else null
                                val protocolIdentifier = jsObject.getString(Param.PROTOCOL_IDENTIFIER)
                                val networkId = jsObject.getString(Param.NETWORK_ID)
                                val moduleIdentifier = jsObject.getString(Param.MODULE_IDENTIFIER)

                                val value = when (target) {
                                    JSCallMethodTarget.OfflineProtocol -> {
                                        assertReceivedIn(jsObject, Param.PROTOCOL_IDENTIFIER)
                                        jsEvaluator.evaluateCallOfflineProtocolMethod(method, args, protocolIdentifier!!, runRef)
                                    }
                                    JSCallMethodTarget.OnlineProtocol -> {
                                        assertReceivedIn(jsObject, Param.PROTOCOL_IDENTIFIER)
                                        jsEvaluator.evaluateCallOnlineProtocolMethod(method, args, protocolIdentifier!!, networkId, runRef)
                                    }
                                    JSCallMethodTarget.BlockExplorer -> {
                                        assertReceivedIn(jsObject, Param.PROTOCOL_IDENTIFIER)
                                        jsEvaluator.evaluateCallBlockExplorerMethod(method, args, protocolIdentifier!!, networkId, runRef)
                                    }
                                    JSCallMethodTarget.V3SerializerCompanion -> {
                                        assertReceivedIn(jsObject, Param.MODULE_IDENTIFIER)
                                        jsEvaluator.evaluateCallV3SerializerCompanionMethod(method, args, moduleIdentifier!!, runRef)
                                    }
                                }

                                JSObject("""
                                    {
                                        "type": "success",
                                        "value": ${value.get("value").serialize()}
                                    }
                                """.trimIndent())
                            } catch (e: Exception) {
                                JSObject("""
                                    {
                                        "type": "error",
                                        "error": "${e.message}"
                                    }
                                """.trimIndent())
                            }
                        }
                    }

                    resolve(JSObject("""
                        {
                            "values": ${JSArray(values)}
                        }
                    """.trimMargin()))
                }
            }
        }
    }

    override fun handleOnDestroy() {
        super.handleOnDestroy()
        activity.lifecycleScope.launch {
            jsEvaluator.await().destroy()
        }
    }

    private val PluginCall.path: String
        get() = getString(Param.PATH)!!

    private val PluginCall.directory: Directory
        get() = getString(Param.DIRECTORY)?.let { Directory.fromString(it) }!!

    private val PluginCall.identifier: String
        get() = getString(Param.IDENTIFIER)!!

    private val PluginCall.identifiers: List<String>?
        get() = getArray(Param.IDENTIFIERS, null)?.toList()

    private  val PluginCall.protocolIdentifiers: List<String>
        get() = getArray(Param.PROTOCOL_IDENTIFIERS).toList()

    private val PluginCall.protocolType: JSProtocolType?
        get() = getString(Param.PROTOCOL_TYPE)?.let { JSProtocolType.fromString(it) }

    private val PluginCall.ignoreProtocols: JSArray?
        get() = getArray(Param.IGNORE_PROTOCOLS, null)

    private val PluginCall.target: JSCallMethodTarget
        get() = getString(Param.TARGET)?.let { JSCallMethodTarget.fromString(it) }!!

    private val PluginCall.method: String
        get() = getString(Param.METHOD)!!

    private val PluginCall.args: JSArray?
        get() = getArray(Param.ARGS, null)

    private val PluginCall.protocolIdentifier: String
        get() = getString(Param.PROTOCOL_IDENTIFIER)!!

    private val PluginCall.moduleIdentifier: String
        get() = getString(Param.MODULE_IDENTIFIER)!!

    private val PluginCall.networkId: String?
        get() = getString(Param.NETWORK_ID)

    private val PluginCall.options: JSArray
        get() = getArray(Param.OPTIONS)

    private object Param {
        const val PATH = "path"
        const val DIRECTORY = "directory"
        const val IDENTIFIER = "identifier"
        const val IDENTIFIERS = "identifiers"
        const val PROTOCOL_IDENTIFIERS = "protocolIdentifiers"
        const val PROTOCOL_TYPE = "protocolType"
        const val IGNORE_PROTOCOLS = "ignoreProtocols"
        const val TARGET = "target"
        const val METHOD = "method"
        const val ARGS = "args"
        const val PROTOCOL_IDENTIFIER = "protocolIdentifier"
        const val MODULE_IDENTIFIER = "moduleIdentifier"
        const val NETWORK_ID = "networkId"
        const val OPTIONS = "options"
    }
}